edb/lib/ext/auth.edgeql

```diff
@@ -48,4 +48,114 @@ CREATE EXTENSION PACKAGE auth VERSION '1.0' {
             set default := '00000000000000000000000000000000';
         };
     };
+
+    create scalar type ext::auth::JWTAlgo extending enum<RS256, HS256>;
+
+    create function ext::auth::_jwt_check_signature(
+        jwt: tuple<header: std::str, payload: std::str, signature: std::str>,
+        key: std::str,
+        algo: ext::auth::JWTAlgo = ext::auth::JWTAlgo.RS256,
+    ) -> std::json
+    {
+        set volatility := 'Volatile';  # work-around due to assert()
+        using (
+            with
+                module ext::auth,
+                msg := jwt.header ++ "." ++ jwt.payload,
+                hash := (
+                    "sha256" if algo = JWTAlgo.RS256 or algo = JWTAlgo.HS256
+                    else <str>std::assert(
+                        false, message := "unsupported JWT algo")
+                ),
+            select
+                std::to_json(
+                    std::to_str(
+                        std::enc::base64_decode(
+                            jwt.payload,
+                            padding := false,
+                            alphabet := std::enc::Base64Alphabet.urlsafe,
+                        ),
+                    ),
+                )
+            order by
+                assert(
+                    std::enc::base64_encode(
+                        ext::pgcrypto::hmac(msg, key, hash),
+                        padding := false,
+                        alphabet := std::enc::Base64Alphabet.urlsafe,
+                    ) = jwt.signature,
+                    message := "JWT signature mismatch",
+                )
```

Comment:
Out of curiosity, why are the asserts in the `order by` clause?

Reply:
It makes sure the assertion only happens on objects that pass your filter.

Reply:
Semantically it makes more sense to put assertions in `filter`, but that affects result cardinality inference, so it's in `order by`.  Assertions need to be placed in a way that makes sure they actually get executed (just putting them in a `with` and not using the alias will not work).